### PR TITLE
Add packing functionality 

### PR DESCRIPTION
**Pull request by Cohedrin**

### CODE REVIEW

File: ext/capn_proto/dynamic_struct_builder.cc [hunk 1]
```diff
@@ -17,7 +17,8 @@ namespace ruby_capn_proto {
     ClassBuilder("DynamicStructBuilder", rb_cObject).
       defineAlloc(&alloc).
       defineMethod("which", &which).
-      defineMethod("write", &write).
+      defineMethod("write", &writePacked).
+      defineMethod("writePacked", &writePacked).
```
Comment (jwilm) [inline ext/capn_proto/dynamic_struct_builder.cc]: Ruby method should be 🐍 case

File: ext/capn_proto/dynamic_struct_builder.cc [hunk 2]
```diff
@@ -17,7 +17,8 @@ namespace ruby_capn_proto {
     ClassBuilder("DynamicStructBuilder", rb_cObject).
       defineAlloc(&alloc).
       defineMethod("which", &which).
-      defineMethod("write", &write).
+      defineMethod("write", &writePacked).
```
Comment (jwilm) [inline ext/capn_proto/dynamic_struct_builder.cc]: This seems like it should point to the write method not writePacked.

**Review by jwilm — COMMENTED** [review]
Couple of bits but otherwise looks good. Why add the dockerfile?

File: ext/capn_proto/dynamic_struct_builder.cc (hunk 2 above)
Comment (Cohedrin) [reply ext/capn_proto/dynamic_struct_builder.cc]: My mistake, was cleaning up after making it work and commited this wrong.

**Comment by Cohedrin** [comment]
Dockerfile was added just to make development easier. It's sort of a pain to setup since the gem requires some specific dependencies and bundler will only use the default g++ compiler (so if the global default is not g++7, its not going to build). There may be a working way to specify it, but both @pbchekin and I had trouble getting it working.

I can definitely remove it, just figured it would make other people's (including myself)'s work on it easier.

File: ext/capn_proto/dynamic_struct_builder.cc (hunk 1 above)
Comment (Cohedrin) [reply ext/capn_proto/dynamic_struct_builder.cc]: updated to both be snake case. Rest of the (gem) code base uses snake case for method names as well.

**Review by jwilm — APPROVED** [review]
No need to remove it, I was just curious if it was for dev purposes or something else.

LGTM.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
+FROM ubuntu:16.04
```
Comment (pbchekin) [inline Dockerfile:1]: You can use `ruby:2.4` as a base image and skip ruby and bundler installation steps

File: ext/capn_proto/dynamic_struct_builder.cc [hunk 4]
```diff
@@ -148,7 +149,24 @@ namespace ruby_capn_proto {
     }
   }
 
-  VALUE DynamicStructBuilder::to_bytes(VALUE self) {
+  VALUE DynamicStructBuilder::write_packed(VALUE self, VALUE file) {
+    VALUE rb_fileno = rb_funcall(file, rb_intern("fileno"), 0);
+    int fileno = FIX2INT(rb_fileno);
+    if (!RTEST(rb_iv_get(self, "is_root"))) {
+      rb_raise(Exception::Class, "You can only call wirte_packed() on the message's root struct.");
```
Comment (pbchekin) [inline ext/capn_proto/dynamic_struct_builder.cc]: `wirte_packed` - typo

**Comment by pbchekin** [comment]
The binary file `lib/capn_proto/capn_proto.so` shouldn't be under version control

**Comment by jwilm** [comment]
Given that the binary file was added to git in this branch, we'll want to squash it out before merging.

**Pull request merged**